In [153]:
import pandas as pd
import re

In [154]:
patients=pd.read_csv('patients.csv')
treatment=pd.read_csv('treatments.csv')
adverse=pd.read_csv('adverse_reactions.csv')
treatment_cut=pd.read_csv('treatments_cut.csv')

In [155]:
print(patients.shape)
print(treatment.shape)
print(adverse.shape)
print(treatment_cut.shape)

(503, 14)
(280, 7)
(34, 3)
(70, 7)


In [156]:
treatment_cut

given_name        surname    auralin    novodra  hba1c_start  hba1c_end  \
0         jožka      resanovič  22u - 30u          -         7.56       7.22   
1     inunnguaq       heilmann  57u - 67u          -         7.85       7.45   
2         alwin       svensson  36u - 39u          -         7.78       7.34   
3           thể          lương          -  61u - 64u         7.64       7.22   
4        amanda        ribeiro  36u - 44u          -         7.85       7.47   
5          meya        sjöberg          -  42u - 42u         7.77       7.42   
6          sara          miles  30u - 35u          -         7.53       7.12   
7       frydryk        adamski  63u - 74u          -         7.75       7.27   
8          jane        citizen          -  37u - 38u         7.98       7.60   
9        angela     lavrentyev          -  28u - 24u         7.61       7.14   
10       edelma    villalpando          -  24u - 26u         7.99       7.56   
11     gabriela   michalčíková  26u - 36u          -         8.37       7.97   
12       annika          vaara          -  20u - 21u         7.73       7.34   
13     bernarda        cindrić  40u - 49u          -         7.89       7.55   
14        chiho           higa          -  46u - 46u         7.71       7.30   
15         dani         antoun  36u - 44u          -         7.73       7.34   
16       mijael         guerra  37u - 47u          -         7.85       7.46   
17         emma         barlow  48u - 57u          -         7.85       7.39   
18       flavia     fiorentino  46u - 57u          -         7.95       7.60   
19          noe         aranda  26u - 34u          -         7.51       7.17   
20     seweryna     czerwinska  39u - 46u          -         7.66       7.30   
21        david      villadsen  26u - 39u          -         7.86       7.40   
22    beatrycze        woźniak          -  26u - 27u         7.54       7.17   
23     miłosław     wiśniewski          -  34u - 33u         7.51       7.08   
24          lóa   hrafnsdóttir  22u - 31u          -         7.60       7.09   
25    siebrigje      koldenhof  30u - 38u          -         7.90       7.59   
26      firenze          fodor          -  30u - 35u         7.89       7.55   
27          zoe        wellish          -  33u - 33u         7.71       7.30   
28          una  traustadóttir          -  35u - 34u         8.00       7.50   
29        luboš          pecha          -  30u - 27u         7.79       7.45   
..          ...            ...        ...        ...          ...        ...   
40      eufemio        rosario          -  37u - 40u         7.54       7.26   
41     dalmacia         madrid          -  26u - 23u         7.67       7.21   
42        daimy          tromp          -  40u - 45u         9.41       8.94   
43   hans peter       radislav  42u - 53u          -         7.60       7.20   
44       jeremy        montagu          -  52u - 52u         7.68       7.36   
45    štefanija        novosel  37u - 42u          -         7.64       7.17   
46         anja        mueller  43u - 56u          -         7.58       7.29   
47       regolo          nucci  51u - 59u          -         7.53       7.02   
48      nebechi      ekechukwu          -  37u - 39u         7.78       7.39   
49      satsita      batukayev          -  42u - 42u         7.63       7.25   
50      timothy         cotton          -  26u - 25u         7.92       7.52   
51      bjørnar         nilsen          -  36u - 33u         7.99       7.70   
52       vivien         zsinkó  30u - 38u          -         8.35       7.92   
53        genet       fesahaye  25u - 35u          -         7.73       7.36   
54        borna      šlezinger          -  42u - 41u         7.55       7.18   
55         mary          adams          -  32u - 33u         7.65       7.26   
56       blanka       jurković  31u - 43u          -         7.77       7.33   
57       joseph         tucker  48u - 56u          -         7.67       7.30   
58  c

In [157]:
result = pd.concat([treatment, treatment_cut], ignore_index=True, sort=False)

In [158]:
result

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       veronika      jindrová  41u - 48u          -         7.63       7.20   
1         elliot    richardson          -  40u - 45u         7.56       7.09   
2       yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3           skye    gormanston  33u - 36u          -         7.97       7.62   
4         alissa        montez          -  33u - 29u         7.78       7.46   
5        jasmine         sykes          -  42u - 44u         7.56       7.18   
6         sophia        haugen  37u - 42u          -         7.65       7.27   
7          eddie        archer  31u - 38u          -         7.89       7.55   
8          saber        ménard          -  54u - 54u         8.08       7.70   
9           asia       woźniak  30u - 36u          -         7.76       7.37   
10        joseph           day  29u - 36u          -         7.70       7.19   
11     kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12       roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13        gregor          bole          -  47u - 45u         7.61       7.16   
14        simone  baumgaertner  27u - 37u          -         7.74       7.30   
15          enco        žibrik  55u - 68u          -         7.78       7.34   
16       camilla      zaitseva  28u - 37u          -         7.53       7.13   
17          gina          cain          -  36u - 36u         7.88       7.40   
18    addolorata      lombardi          -  49u - 46u         7.75       7.33   
19        khalid      johnsrud          -  54u - 54u         8.35       7.94   
20          mile        stanić          -  47u - 48u         7.66       7.24   
21         tekla       walczak  29u - 39u          -         7.61       7.29   
22   brancaleone         russo  53u - 60u          -         8.61       8.18   
23      chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24          isac          berg  31u - 41u          -         9.68       9.29   
25        benoît        bonami          -  44u - 43u         9.82       9.40   
26        suhaim         rahal          -  49u - 47u         7.94       7.50   
27        mizuki         iwata          -  45u - 46u         7.70       7.23   
28       clinton        miller  42u - 51u          -         7.79       7.40   
29        eugene       mironov  42u - 49u          -         7.81       7.48   
..           ...           ...        ...        ...          ...        ...   
320      eufemio       rosario          -  37u - 40u         7.54       7.26   
321     dalmacia        madrid          -  26u - 23u         7.67       7.21   
322        daimy         tromp          -  40u - 45u         9.41       8.94   
323   hans peter      radislav  42u - 53u          -         7.60       7.20   
324       jeremy       montagu          -  52u - 52u         7.68       7.36   
325    štefanija       novosel  37u - 42u          -         7.64       7.17   
326         anja       mueller  43u - 56u          -         7.58       7.29   
327       regolo         nucci  51u - 59u          -         7.53       7.02   
328      nebechi     ekechukwu          -  37u - 39u         7.78       7.39   
329      satsita     batukayev          -  42u - 42u         7.63       7.25   
330      timothy        cotton          -  26u - 25u         7.92       7.52   
331      bjørnar        nilsen          -  36u - 33u         7.99       7.70   
332       vivien        zsinkó  30u - 38u          -         8.35       7.92   
333        genet      fesahaye  25u - 35u          -         7.73       7.36   
334        borna     šlezinger          -  42u - 41u         7.55       7.18   
335         mary         adams          -  32u - 33u         7.65       7.26   
336       blanka      jurković  31u - 43u          -         7.77       7.33   
337       joseph        tucker  48u - 56u          -         7.67       7.30   
338  c

In [159]:
patients

patient_id assigned_sex  given_name       surname  \
0             1       female         Zoe       Wellish   
1             2       female      Pamela          Hill   
2             3         male         Jae        Debord   
3             4         male        Liêm          Phan   
4             5         male         Tim       Neudorf   
5             6         male      Rafael         Costa   
6             7       female        Mary         Adams   
7             8       female      Xiuxiu         Chang   
8             9         male       Dsvid    Gustafsson   
9            10       female      Sophie       Cabrera   
10           11       female       Sandy    Gunnarsson   
11           12         male   Abdul-Nur           Isa   
12           13         male  Omeokachie      Ibeamaka   
13           14       female    Anenechi         Chidi   
14           15       female        Asia       Woźniak   
15           16         male       Søren          Lund   
16           17       female         Tám          Liễu   
17           18       female     Roxanne     Andreyeva   
18           19         male     William         Oates   
19           20         male         Zak         Kelly   
20           21       female       Sofia       Karlsen   
21           22         male      Samúel  Guðbrandsson   
22           23         male      Manchu            Su   
23           24         male       Lovre         Galić   
24           25         male       Jakob      Jakobsen   
25           26         male      Gregor          Bole   
26           27       female        Ella          Lund   
27           28         male      Joseph        Tucker   
28           29         male      Robert          Wolf   
29           30         male        Jake      Jakobsen   
..          ...          ...         ...           ...   
473         474       female        Kate     Wilkinson   
474         475       female   Esperanza      Labrosse   
475         476         male       Malik       Vaneker   
476         477       female       Berta    Napolitani   
477         478         male     Juliusz      Majewski   
478         479       female      Edelma   Villalpando   
479         480         male        Tapa   Arsanukayev   
480         481         male      Nasser       Mansour   
481         482         male     Michael    Kristensen   
482         483         male       Diogo         Souza   
483         484       female       Angel         Grant   
484         485         male     Placido       Udinesi   
485         486         male      Trifon      Izmailov   
486         487         male      Samuel          Blix   
487         488         male        Ivar       Löfgren   
488         489         male        Mika    Martinsson   
489         490       female     Jasmine         Sykes   
490         491         male     Jackson       Addison   
491         492       female     Vanessa      Ferguson   
492         493         male       Poldi           Tar   
493         494       female         Fen          Chin   
494         495       female      Sirkka     Piirainen   
495         496         male      Hajime       Tsukada   
496         497         male   Alexander        Hueber   
497         498         male    Masataka      Murakami   
498         499         male     Mustafa     Lindström   
499         500         male       Ruman       Bisliev   
500         501       female       Jinke     de Keizer   
501         502       female     Chidalu   Onyekaozulu   
502         503         male         Pat       Gersten   

                        address               city       state  zip_code  \
0          576 Brown Bear Drive  Rancho California  California   92390.0   
1     2370 University Hill Road          Armstrong    Illinois   61812.0   
2         1493 Poling Farm Road               York    Nebraska   68467.0   
3           2335 Webster Street         Woodbridge          NJ    7095.0   
4          1428 Turkey Pen L

In [160]:
treatment

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       veronika      jindrová  41u - 48u          -         7.63       7.20   
1         elliot    richardson          -  40u - 45u         7.56       7.09   
2       yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3           skye    gormanston  33u - 36u          -         7.97       7.62   
4         alissa        montez          -  33u - 29u         7.78       7.46   
5        jasmine         sykes          -  42u - 44u         7.56       7.18   
6         sophia        haugen  37u - 42u          -         7.65       7.27   
7          eddie        archer  31u - 38u          -         7.89       7.55   
8          saber        ménard          -  54u - 54u         8.08       7.70   
9           asia       woźniak  30u - 36u          -         7.76       7.37   
10        joseph           day  29u - 36u          -         7.70       7.19   
11     kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12       roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13        gregor          bole          -  47u - 45u         7.61       7.16   
14        simone  baumgaertner  27u - 37u          -         7.74       7.30   
15          enco        žibrik  55u - 68u          -         7.78       7.34   
16       camilla      zaitseva  28u - 37u          -         7.53       7.13   
17          gina          cain          -  36u - 36u         7.88       7.40   
18    addolorata      lombardi          -  49u - 46u         7.75       7.33   
19        khalid      johnsrud          -  54u - 54u         8.35       7.94   
20          mile        stanić          -  47u - 48u         7.66       7.24   
21         tekla       walczak  29u - 39u          -         7.61       7.29   
22   brancaleone         russo  53u - 60u          -         8.61       8.18   
23      chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24          isac          berg  31u - 41u          -         9.68       9.29   
25        benoît        bonami          -  44u - 43u         9.82       9.40   
26        suhaim         rahal          -  49u - 47u         7.94       7.50   
27        mizuki         iwata          -  45u - 46u         7.70       7.23   
28       clinton        miller  42u - 51u          -         7.79       7.40   
29        eugene       mironov  42u - 49u          -         7.81       7.48   
..           ...           ...        ...        ...          ...        ...   
250         chen           yao          -  56u - 57u         7.90       7.51   
251        aksel   vestergaard          -  42u - 38u         9.62       9.29   
252        ellen         luman          -  40u - 39u         9.27       8.77   
253       albino     schiavone  35u - 43u          -         7.56       7.15   
254         jose         combs          -  39u - 36u         7.89       7.42   
255       jia li          teng  48u - 54u          -         7.66       7.32   
256        ilija        horvat  42u - 50u          -         7.77       7.38   
257     mathilde      nørgaard          -  27u - 28u         8.50       8.10   
258       csilla      herczegh          -  43u - 46u         7.71       7.27   
259      aaliyah          rice          -  31u - 31u         7.64       7.33   
260        david      beauvais          -  26u - 23u         7.87       7.47   
261     caroline        shuler          -  50u - 54u         7.63       7.27   
262         alex      crawford  51u - 62u          -         7.69       7.30   
263      rebecca      jephcott  53u - 63u          -         7.96       7.57   
264   chukwumoge     ogochukwu          -  41u - 39u         7.95       7.56   
265       fearne      mcgregor          -  27u - 29u         7.83       7.48   
266       ursula         freud  42u - 54u          -         7.75       7.46   
267         leon        scholz          -  38u - 32u         7.72       7.29   
268   

In [161]:
adverse

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [162]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [163]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [164]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [165]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [166]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [167]:
treatment.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [168]:
patients.sample(5)

patient_id assigned_sex given_name    surname                 address  \
47           48       female        Ana    Correia     1495 Post Farm Road   
412         413         male     Finlay   Sheppard  4571 Walnut Hill Drive   
154         155       female       Maya   Isaksson      1333 Comfort Court   
295         296         male     Albino  Schiavone  3113 Timber Ridge Road   
97           98         male    Patrick    Gersten       2778 North Avenue   

           city       state  zip_code        country  \
47      Atlanta          GA   30337.0  United States   
412    Portland          OH   45204.0  United States   
154     Madison          WI   53711.0  United States   
295  Sacramento  California   95814.0  United States   
97         Burr          NE   68324.0  United States   

                                         contact  birthdate  weight  height  \
47         404-289-1766AnaSousaCorreia@rhyta.com  11/4/1962   186.6      65   
412  FinlaySheppard@fleckens.hu+1 (513) 297-6359  6/16/1948   161.3      70   
154    MayaIsaksson@jourrapide.com1 608 277 0146  7/14/1959   137.7      67   
295    AlbinoSchiavone@fleckens.hu1 916 817 9960  6/30/1928   180.0      67   
97          PatrickGersten@rhyta.com402-848-4923   5/3/1954   138.2      71   

      bmi  
47   31.0  
412  23.1  
154  21.6  
295  28.2  
97   19.3

In [169]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

### Quality

[Completeness, Validity, Accuracy, Consistency]

#### `Patients`

- Zip code col is stored in decimal format[V]
- Certain zip codes of 4 digits[V] 
- For the state col sometimes full names are use and other times short forms are used[IC]
- Missing values for contact details of 12 patients[C]
- Incorrect data types for assigned_sex and birthdate cols[?]
- Erroneous data for weights and heights[Accuracy]
- Inconsistent way of storing phone numbers in contact col[IC]
- Duplicate and not useful data in dataset[IC]

#### `Treatments`

- hba1c_change col has missing values[C]
- Novodra and Auralin cols has u attached to numbers[V]
- Only 280 rows are there, instead of 350[C]
- Name and surname col is in lower case[IC]
- Incorrect data type for auralin and novadra cols[V]
- hba1c_change 9 instead of 4[Accuracy]
- use of dashes instead of Nan in aurolin and novodra cols[IC]

#### `Adverse Reactions`

- Name and surname col is in lower case[IC]

### Tidiness

#### `Patients`

- Phone and email together in one col i.e. contact

#### `Treatments`

- 3 variables stored in auralin and novodra cols

#### `Adverse Reactions`

- No point of using this table


# Cleaning Data

You should follow the order provided below:

- Completeness Issues
- Tidiness Issues
- Remaining Data quality issues like validity, accuracy and consistency

### Steps involved in Data cleaning

- Define
- Code 
- Test

`Always make sure to create a copy of your pandas dataframe before you start the cleaning process`

In [170]:
# Creating copies

patients_copy=patients.copy()
treatment_copy=treatment.copy()
adverse_copy=adverse.copy()
treatment_cut_copy=treatment_cut.copy()

In [171]:
patients_copy[patients_copy['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

## a. Completeness Issues

- Missing values for contact details of 12 patients
- Only 280 rows are there, instead of 350
- hba1c_change col has missing values

#### Define

In [172]:
# Merge Treatment_copy and Treatment_cut

In [173]:
# Subtract hba1c_start from hba1c_end to get hba1c_change

#### Code

In [174]:
treatments_final=pd.concat([treatment_copy, treatment_cut_copy], ignore_index=True, sort=False)

In [175]:
treatments_final

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       veronika      jindrová  41u - 48u          -         7.63       7.20   
1         elliot    richardson          -  40u - 45u         7.56       7.09   
2       yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3           skye    gormanston  33u - 36u          -         7.97       7.62   
4         alissa        montez          -  33u - 29u         7.78       7.46   
5        jasmine         sykes          -  42u - 44u         7.56       7.18   
6         sophia        haugen  37u - 42u          -         7.65       7.27   
7          eddie        archer  31u - 38u          -         7.89       7.55   
8          saber        ménard          -  54u - 54u         8.08       7.70   
9           asia       woźniak  30u - 36u          -         7.76       7.37   
10        joseph           day  29u - 36u          -         7.70       7.19   
11     kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12       roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13        gregor          bole          -  47u - 45u         7.61       7.16   
14        simone  baumgaertner  27u - 37u          -         7.74       7.30   
15          enco        žibrik  55u - 68u          -         7.78       7.34   
16       camilla      zaitseva  28u - 37u          -         7.53       7.13   
17          gina          cain          -  36u - 36u         7.88       7.40   
18    addolorata      lombardi          -  49u - 46u         7.75       7.33   
19        khalid      johnsrud          -  54u - 54u         8.35       7.94   
20          mile        stanić          -  47u - 48u         7.66       7.24   
21         tekla       walczak  29u - 39u          -         7.61       7.29   
22   brancaleone         russo  53u - 60u          -         8.61       8.18   
23      chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24          isac          berg  31u - 41u          -         9.68       9.29   
25        benoît        bonami          -  44u - 43u         9.82       9.40   
26        suhaim         rahal          -  49u - 47u         7.94       7.50   
27        mizuki         iwata          -  45u - 46u         7.70       7.23   
28       clinton        miller  42u - 51u          -         7.79       7.40   
29        eugene       mironov  42u - 49u          -         7.81       7.48   
..           ...           ...        ...        ...          ...        ...   
320      eufemio       rosario          -  37u - 40u         7.54       7.26   
321     dalmacia        madrid          -  26u - 23u         7.67       7.21   
322        daimy         tromp          -  40u - 45u         9.41       8.94   
323   hans peter      radislav  42u - 53u          -         7.60       7.20   
324       jeremy       montagu          -  52u - 52u         7.68       7.36   
325    štefanija       novosel  37u - 42u          -         7.64       7.17   
326         anja       mueller  43u - 56u          -         7.58       7.29   
327       regolo         nucci  51u - 59u          -         7.53       7.02   
328      nebechi     ekechukwu          -  37u - 39u         7.78       7.39   
329      satsita     batukayev          -  42u - 42u         7.63       7.25   
330      timothy        cotton          -  26u - 25u         7.92       7.52   
331      bjørnar        nilsen          -  36u - 33u         7.99       7.70   
332       vivien        zsinkó  30u - 38u          -         8.35       7.92   
333        genet      fesahaye  25u - 35u          -         7.73       7.36   
334        borna     šlezinger          -  42u - 41u         7.55       7.18   
335         mary         adams          -  32u - 33u         7.65       7.26   
336       blanka      jurković  31u - 43u          -         7.77       7.33   
337       joseph        tucker  48u - 56u          -         7.67       7.30   
338  c

In [176]:
treatments_final['hba1c_change']=treatments_final['hba1c_start']-treatments_final['hba1c_end']

#### Test

In [177]:
treatments_final.info()

treatments_final.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    350 non-null float64
dtypes: float64(3), object(4)
memory usage: 19.2+ KB


hba1c_start   hba1c_end  hba1c_change
count   350.000000  350.000000    350.000000
mean      7.956343    7.560057      0.396286
std       0.545328    0.545456      0.059283
min       7.500000    7.010000      0.200000
25%       7.650000    7.270000      0.350000
50%       7.785000    7.400000      0.390000
75%       7.950000    7.557500      0.440000
max       9.950000    9.580000      0.540000

## b. Tidiness Issues

- Phone + email in contact col `patients`
- Novodra and Aurolin col `treatment`
- Merge the `adverse` table to `treatment` table

#### Define

In [178]:
# Phone and email in separate cols

In [179]:
# Merge the adverse table to treatment table

In [180]:
# Novodra and Aurolin using the melt function

#### Code

In [181]:
patients_1=patients_copy[patients_copy['address'].isnull()]
patients_copy.dropna(inplace=True)

In [182]:
emails=[]
numbers=[]
for i in patients_copy['contact']:
    pattern=re.compile(r'[a-zA-Z][a-zA-Z0-9.-]+[@][a-zA-Z0-9-]+\.[a-zA-Z0-9.-]+[a-zA-Z]')
    numbers.append(i.replace(pattern.findall(i)[0],''))
    emails.append(pattern.findall(i)[0])

In [183]:
patients_copy['Emails']=emails
patients_copy['Phone numbers']=numbers

patients_copy=pd.concat([patients_copy,patients_1]).reset_index()

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  after removing the cwd from sys.path.


In [184]:
patients_copy=patients_copy.drop(columns=['index','contact'])

In [185]:
patients_copy

Emails      Phone numbers  \
0             ZoeWellish@superrito.com       951-719-9170   
1                 PamelaSHill@cuvox.de  +1 (217) 569-3204   
2                 JaeMDebord@gustr.com       402-363-6804   
3            PhanBaLiem@jourrapide.com  +1 (732) 636-8246   
4                  TimNeudorf@cuvox.de       334-515-7487   
5         RafaelCardosoCosta@gustr.com       386-334-5237   
6                MaryBAdams@einrot.com       775-533-5933   
7               XiuxiuChang@einrot.com     1 408 778 3236   
8          DavidGustafsson@armyspy.com       816-265-9578   
9      SophieCabreraIbarra@teleworm.us     1 718 795 9124   
10          SandyGunnarsson@dayrep.com       906-478-8949   
11        Abdul-NurMummarIsa@rhyta.com     1 931 207 0839   
12       OmeokachieIbeamaka@einrot.com       434-509-2614   
13           AnenechiChidi@armyspy.com  +1 (205) 417-8095   
14               AsiaWozniak@rhyta.com       918-712-3469   
15                 SrenFLund@gustr.com       276-225-1955   
16            LieuThiThuTam@dayrep.com     1 559 765 7836   
17        RoxanneAndreyeva@armyspy.com       678-829-8578   
18           WilliamVOates@armyspy.com       406-775-2696   
19                  ZakKelly@rhyta.com     1 530 532 8397   
20           SofiaTKarlsen@teleworm.us     1 781 447 1763   
21       SamuelGubrandsson@teleworm.us       973-445-5341   
22                 ManchuSu@einrot.com       914-745-6108   
23                LovreGalic@gustr.com     1 813 355 9476   
24           JakobCJakobsen@einrot.com  +1 (845) 858-7707   
25                GregorBole@gustr.com       505-828-4955   
26                EllaLund@armyspy.com       309-671-8852   
27             JosephNTucker@rhyta.com       985-814-7603   
28              RobertWolf@fleckens.hu     1 267 895 7462   
29           JakobCJakobsen@einrot.com  +1 (845) 858-7707   
..                                 ...                ...   
473         TrifonIzmailov@fleckens.hu     1 850 659 0417   
474              SamuelBlix@dayrep.com       254-681-4504   
475            IvarLofgren@armyspy.com     1 785 229 1188   
476         MikaMartinsson@armyspy.com       352-453-4601   
477        JasmineSykes@jourrapide.com       925-283-5425   
478         JacksonAddison@armyspy.com       914-636-9304   
479     VanessaFerguson@jourrapide.com       210-222-8684   
480             TarPoldi@superrito.com       714-496-2264   
481                  FenChin@gustr.com  +1 (208) 388-1065   
482        SirkkaPiirainen@teleworm.us  +1 (586) 790-0975   
483           HajimeTsukada@dayrep.com       650-570-4896   
484     AlexanderHueber@jourrapide.com     1 209 762 2320   
485        MasatakaMurakami@einrot.com  +1 (918) 984-9171   
486    MustafaLindstrom@jourrapide.com       207-477-0579   
487             RumanBisliev@gustr.com       928-284-4492   
488          JinkedeKeizer@teleworm.us       816-223-6007   
489  ChidaluOnyekaozulu@jourrapide.com     1 360 443 2060   
490           PatrickGersten@rhyta.com       402-848-4923   
491                                NaN                NaN   
492                                NaN                NaN   
493                                NaN                NaN   
494                                NaN                NaN   
495                                NaN                NaN   
496                                NaN                NaN   
497                                NaN                NaN   
498                                NaN                NaN   
499                                NaN                NaN   
500                                NaN                NaN   
501                                NaN                NaN   
502                                NaN                NaN   

                        address assigned_sex   birthdate   bmi  \
0          576 Brown Bear Drive       female   7/10/1976  19.6   
1     2370 University Hill Road       female    4/3/1967  19.2   
2         1493 Poling Farm Road         male   2/19

In [186]:
treatments_final=pd.melt(treatments_final, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')

In [187]:
treatments_final=treatments_final[treatments_final['dose']!='-']

In [188]:
treatments_final['start'],treatments_final['end']=treatments_final['dose'].str.split('-').str

In [189]:
treatments_final.drop(columns={'dose'}, inplace=True)

In [190]:
treatments_final['start']=treatments_final['start'].str.replace('u','')
treatments_final['end']=treatments_final['end'].str.replace('u','')

treatments_final['start']=treatments_final['start'].astype('int')
treatments_final['end']=treatments_final['end'].astype('int')

treatments_final.reset_index(drop=True,inplace=True)

#### Test

In [191]:
treatments_final.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1       skye  gormanston         7.97       7.62          0.35   auralin   
2     sophia      haugen         7.65       7.27          0.38   auralin   
3      eddie      archer         7.89       7.55          0.34   auralin   
4       asia     woźniak         7.76       7.37          0.39   auralin   

   start  end  
0     41   48  
1     33   36  
2     37   42  
3     31   38  
4     30   36

In [192]:
treatments_final=treatments_final.merge(adverse_copy,how='left',on=['given_name','surname'],)

In [193]:
treatments_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 9 columns):
given_name          350 non-null object
surname             350 non-null object
hba1c_start         350 non-null float64
hba1c_end           350 non-null float64
hba1c_change        350 non-null float64
treatment           350 non-null object
start               350 non-null int32
end                 350 non-null int32
adverse_reaction    35 non-null object
dtypes: float64(3), int32(2), object(4)
memory usage: 24.6+ KB


## c. Validity Issues
- zipcode in 4 digits in some places 
- zipcode in decimal format

In [194]:
patients_copy['zip_code']=patients_copy['zip_code'].astype('str')

patients_copy['zip_code']=patients_copy['zip_code'].str.replace('.0','',regex=False).str.pad(width=5,side='left',fillchar='0')


patients_copy['zip_code']

0      92390
1      61812
2      68467
3      07095
4      36303
5      32114
6      84728
7      95037
8      64105
9      10011
10     49780
11     37027
12     24504
13     35203
14     74105
15     37660
16     93706
17     30082
18     59324
19     95966
20     02382
21     10523
22     10570
23     33830
24     12771
25     87109
26     61602
27     70091
28     19034
29     12771
       ...  
473    32548
474    76706
475    66067
476    34471
477    94549
478    10801
479    78205
480    91505
481    83702
482    48066
483    94403
484    95204
485    74116
486    03852
487    86341
488    64110
489    98109
490    68324
491    00nan
492    00nan
493    00nan
494    00nan
495    00nan
496    00nan
497    00nan
498    00nan
499    00nan
500    00nan
501    00nan
502    00nan
Name: zip_code, Length: 503, dtype: object

In [195]:
patients_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
Emails           491 non-null object
Phone numbers    491 non-null object
address          491 non-null object
assigned_sex     503 non-null object
birthdate        503 non-null object
bmi              503 non-null float64
city             491 non-null object
country          491 non-null object
given_name       503 non-null object
height           503 non-null int64
patient_id       503 non-null int64
state            491 non-null object
surname          503 non-null object
weight           503 non-null float64
zip_code         503 non-null object
dtypes: float64(2), int64(2), object(11)
memory usage: 59.0+ KB


## d. Accuracy issues
- Erronoeus data for height and weight

In [196]:
patients_copy[patients_copy['height']<40]

Emails Phone numbers               address assigned_sex  \
4  TimNeudorf@cuvox.de  334-515-7487  1428 Turkey Pen Lane         male   

   birthdate   bmi    city        country given_name  height  patient_id  \
4  2/18/1928  26.1  Dothan  United States        Tim      27           5   

  state  surname  weight zip_code  
4    AL  Neudorf   192.3    36303

## Code

In [197]:
patients_copy.drop(index=4,inplace=True)

## test

In [198]:
patients_copy.head()

Emails      Phone numbers                    address  \
0      ZoeWellish@superrito.com       951-719-9170       576 Brown Bear Drive   
1          PamelaSHill@cuvox.de  +1 (217) 569-3204  2370 University Hill Road   
2          JaeMDebord@gustr.com       402-363-6804      1493 Poling Farm Road   
3     PhanBaLiem@jourrapide.com  +1 (732) 636-8246        2335 Webster Street   
5  RafaelCardosoCosta@gustr.com       386-334-5237         1140 Willis Avenue   

  assigned_sex  birthdate   bmi               city        country given_name  \
0       female  7/10/1976  19.6  Rancho California  United States        Zoe   
1       female   4/3/1967  19.2          Armstrong  United States     Pamela   
2         male  2/19/1980  24.8               York  United States        Jae   
3         male  7/26/1951  31.7         Woodbridge  United States       Liêm   
5         male  8/31/1931  26.4      Daytona Beach  United States     Rafael   

   height  patient_id       state  surname  weight zip_code  
0      66           1  California  Wellish   121.7    92390  
1      66           2    Illinois     Hill   118.8    61812  
2      71           3    Nebraska   Debord   177.8    68467  
3      70           4          NJ     Phan   220.9    07095  
5      70           6     Florida    Costa   183.9    32114

## e. Inconsistency Issues
- For the state col sometimes full names are use and other times short forms are used
- Duplicate and not useful data in dataset
- Treatments name in small

# Code

In [199]:
treatments_final['given_name']=treatments_final['given_name'].str.capitalize()
treatments_final['surname']=treatments_final['surname'].str.capitalize()

In [200]:
treatments_final.drop_duplicates(inplace=True)
patients_copy.drop_duplicates(inplace=True)

In [201]:
patients_copy = patients_copy[["patient_id","given_name","surname","address","assigned_sex","birthdate","Emails","Phone numbers","height","bmi","weight","city","state","country","zip_code"]]



# #Test

In [202]:
treatments_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 349
Data columns (total 9 columns):
given_name          349 non-null object
surname             349 non-null object
hba1c_start         349 non-null float64
hba1c_end           349 non-null float64
hba1c_change        349 non-null float64
treatment           349 non-null object
start               349 non-null int32
end                 349 non-null int32
adverse_reaction    34 non-null object
dtypes: float64(3), int32(2), object(4)
memory usage: 24.5+ KB


In [203]:
patients_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 502 entries, 0 to 502
Data columns (total 15 columns):
patient_id       502 non-null int64
given_name       502 non-null object
surname          502 non-null object
address          490 non-null object
assigned_sex     502 non-null object
birthdate        502 non-null object
Emails           490 non-null object
Phone numbers    490 non-null object
height           502 non-null int64
bmi              502 non-null float64
weight           502 non-null float64
city             490 non-null object
state            490 non-null object
country          490 non-null object
zip_code         502 non-null object
dtypes: float64(2), int64(2), object(11)
memory usage: 62.8+ KB


In [204]:
patients_copy

patient_id  given_name       surname                     address  \
0             1         Zoe       Wellish        576 Brown Bear Drive   
1             2      Pamela          Hill   2370 University Hill Road   
2             3         Jae        Debord       1493 Poling Farm Road   
3             4        Liêm          Phan         2335 Webster Street   
5             6      Rafael         Costa          1140 Willis Avenue   
6             7        Mary         Adams            3145 Sheila Lane   
7             8      Xiuxiu         Chang  2687 Black Oak Hollow Road   
8             9       Dsvid    Gustafsson          1790 Nutter Street   
9            10      Sophie       Cabrera           3303 Anmoore Road   
10           11       Sandy    Gunnarsson          87 Wood Duck Drive   
11           12   Abdul-Nur           Isa      1092 Farm Meadow Drive   
12           13  Omeokachie      Ibeamaka          2544 Worley Avenue   
13           14    Anenechi         Chidi            826 Broad Street   
14           15        Asia       Woźniak       4970 Heather Sees Way   
15           16       Søren          Lund      2438 Shady Pines Drive   
16           17         Tám          Liễu          2152 Heritage Road   
17           18     Roxanne     Andreyeva         2103 Edington Drive   
18           19     William         Oates            441 Tibbs Avenue   
19           20         Zak         Kelly    994 Hill Croft Farm Road   
20           21       Sofia       Karlsen          2931 Romano Street   
21           22      Samúel  Guðbrandsson         1904 Granville Lane   
22           23      Manchu            Su             1092 Deans Lane   
23           24       Lovre         Galić           4941 Marion Drive   
24           25       Jakob      Jakobsen           648 Old Dear Lane   
25           26      Gregor          Bole           922 Chapmans Lane   
26           27        Ella          Lund          1207 Garfield Road   
27           28      Joseph        Tucker            4982 Wood Street   
28           29      Robert          Wolf           2386 Linda Street   
29           30        Jake      Jakobsen           648 Old Dear Lane   
30           31      Janina      Bazarova              3259 Roy Alley   
..          ...         ...           ...                         ...   
473         486      Trifon      Izmailov         3697 Drainer Avenue   
474         487      Samuel          Blix           3488 Clair Street   
475         488        Ivar       Löfgren        1346 Nicholas Street   
476         489        Mika    Martinsson           962 George Street   
477         490     Jasmine         Sykes           2607 Water Street   
478         491     Jackson       Addison          1160 Taylor Street   
479         492     Vanessa      Ferguson         241 Freshour Circle   
480         493       Poldi           Tar         3958 Liberty Avenue   
481         494         Fen          Chin      1826 Poplar Chase Lane   
482         495      Sirkka     Piirainen          4102 Ritter Avenue   
483         496      Hajime       Tsukada           4111 Thunder Road   
484         497   Alexander        Hueber            3868 Freed Drive   
485         498    Masataka      Murakami            1179 Patton Lane   
486         499     Mustafa     Lindström         2530 Victoria Court   
487         500       Ruman       Bisliev    494 Clarksburg Park Road   
488         501       Jinke     de Keizer           649 Nutter Street   
489         502     Chidalu   Onyekaozulu    3652 Boone Crockett Lane   
490         503         Pat       Gersten           2778 North Avenue   
491         210      Lalita   Eldarkhanov                         NaN   
492         220          Mỹ         Quynh                         NaN   
493         231   Elisabeth       Knudsen                         NaN   
494         235     Martina     Tománková                         NaN   
495         243        John       O'Brian                      

In [205]:
patients_copy['state'].value_counts()

California    36
TX            32
New York      25
CA            24
NY            22
MA            22
PA            18
GA            15
OH            14
Illinois      14
Florida       13
OK            13
LA            13
MI            13
NJ            12
VA            11
WI            10
IL            10
MS            10
IN             9
FL             9
TN             9
MN             9
NC             8
KY             8
AL             8
WA             8
MO             7
ID             6
NV             6
KS             6
CT             5
SC             5
IA             5
ND             4
CO             4
AZ             4
ME             4
AR             4
RI             4
Nebraska       4
MD             3
WV             3
DE             3
SD             3
OR             3
DC             2
MT             2
VT             2
NE             2
NH             1
NM             1
AK             1
WY             1
Name: state, dtype: int64

# Code 

In [206]:
patients_copy['state'].replace({'California':'CA','Illinois':'IL','Nebraska':'NE','Florida':'FL','New York':'NY'},inplace=True)

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\generic.py:5890: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


# # Test 

In [207]:
patients_copy['state'].value_counts()

CA    60
NY    47
TX    32
IL    24
MA    22
FL    22
PA    18
GA    15
OH    14
LA    13
MI    13
OK    13
NJ    12
VA    11
WI    10
MS    10
IN     9
MN     9
TN     9
WA     8
AL     8
KY     8
NC     8
MO     7
KS     6
ID     6
NE     6
NV     6
CT     5
IA     5
SC     5
AR     4
ND     4
ME     4
AZ     4
RI     4
CO     4
WV     3
DE     3
MD     3
OR     3
SD     3
MT     2
DC     2
VT     2
NH     1
NM     1
WY     1
AK     1
Name: state, dtype: int64

# #Result

In [208]:
patients=patients_copy.copy()
treatments=treatments_final.copy()

In [209]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 502 entries, 0 to 502
Data columns (total 15 columns):
patient_id       502 non-null int64
given_name       502 non-null object
surname          502 non-null object
address          490 non-null object
assigned_sex     502 non-null object
birthdate        502 non-null object
Emails           490 non-null object
Phone numbers    490 non-null object
height           502 non-null int64
bmi              502 non-null float64
weight           502 non-null float64
city             490 non-null object
state            490 non-null object
country          490 non-null object
zip_code         502 non-null object
dtypes: float64(2), int64(2), object(11)
memory usage: 62.8+ KB


In [210]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 349
Data columns (total 9 columns):
given_name          349 non-null object
surname             349 non-null object
hba1c_start         349 non-null float64
hba1c_end           349 non-null float64
hba1c_change        349 non-null float64
treatment           349 non-null object
start               349 non-null int32
end                 349 non-null int32
adverse_reaction    34 non-null object
dtypes: float64(3), int32(2), object(4)
memory usage: 24.5+ KB


In [211]:
patients

patient_id  given_name       surname                     address  \
0             1         Zoe       Wellish        576 Brown Bear Drive   
1             2      Pamela          Hill   2370 University Hill Road   
2             3         Jae        Debord       1493 Poling Farm Road   
3             4        Liêm          Phan         2335 Webster Street   
5             6      Rafael         Costa          1140 Willis Avenue   
6             7        Mary         Adams            3145 Sheila Lane   
7             8      Xiuxiu         Chang  2687 Black Oak Hollow Road   
8             9       Dsvid    Gustafsson          1790 Nutter Street   
9            10      Sophie       Cabrera           3303 Anmoore Road   
10           11       Sandy    Gunnarsson          87 Wood Duck Drive   
11           12   Abdul-Nur           Isa      1092 Farm Meadow Drive   
12           13  Omeokachie      Ibeamaka          2544 Worley Avenue   
13           14    Anenechi         Chidi            826 Broad Street   
14           15        Asia       Woźniak       4970 Heather Sees Way   
15           16       Søren          Lund      2438 Shady Pines Drive   
16           17         Tám          Liễu          2152 Heritage Road   
17           18     Roxanne     Andreyeva         2103 Edington Drive   
18           19     William         Oates            441 Tibbs Avenue   
19           20         Zak         Kelly    994 Hill Croft Farm Road   
20           21       Sofia       Karlsen          2931 Romano Street   
21           22      Samúel  Guðbrandsson         1904 Granville Lane   
22           23      Manchu            Su             1092 Deans Lane   
23           24       Lovre         Galić           4941 Marion Drive   
24           25       Jakob      Jakobsen           648 Old Dear Lane   
25           26      Gregor          Bole           922 Chapmans Lane   
26           27        Ella          Lund          1207 Garfield Road   
27           28      Joseph        Tucker            4982 Wood Street   
28           29      Robert          Wolf           2386 Linda Street   
29           30        Jake      Jakobsen           648 Old Dear Lane   
30           31      Janina      Bazarova              3259 Roy Alley   
..          ...         ...           ...                         ...   
473         486      Trifon      Izmailov         3697 Drainer Avenue   
474         487      Samuel          Blix           3488 Clair Street   
475         488        Ivar       Löfgren        1346 Nicholas Street   
476         489        Mika    Martinsson           962 George Street   
477         490     Jasmine         Sykes           2607 Water Street   
478         491     Jackson       Addison          1160 Taylor Street   
479         492     Vanessa      Ferguson         241 Freshour Circle   
480         493       Poldi           Tar         3958 Liberty Avenue   
481         494         Fen          Chin      1826 Poplar Chase Lane   
482         495      Sirkka     Piirainen          4102 Ritter Avenue   
483         496      Hajime       Tsukada           4111 Thunder Road   
484         497   Alexander        Hueber            3868 Freed Drive   
485         498    Masataka      Murakami            1179 Patton Lane   
486         499     Mustafa     Lindström         2530 Victoria Court   
487         500       Ruman       Bisliev    494 Clarksburg Park Road   
488         501       Jinke     de Keizer           649 Nutter Street   
489         502     Chidalu   Onyekaozulu    3652 Boone Crockett Lane   
490         503         Pat       Gersten           2778 North Avenue   
491         210      Lalita   Eldarkhanov                         NaN   
492         220          Mỹ         Quynh                         NaN   
493         231   Elisabeth       Knudsen                         NaN   
494         235     Martina     Tománková                         NaN   
495         243        John       O'Brian                      

In [212]:
treatments

given_name        surname  hba1c_start  hba1c_end  hba1c_change  \
0        Veronika       Jindrová         7.63       7.20          0.43   
1            Skye     Gormanston         7.97       7.62          0.35   
2          Sophia         Haugen         7.65       7.27          0.38   
3           Eddie         Archer         7.89       7.55          0.34   
4            Asia        Woźniak         7.76       7.37          0.39   
5          Joseph            Day         7.70       7.19          0.51   
6         Roxanne      Andreyeva         9.54       9.14          0.40   
7          Simone   Baumgaertner         7.74       7.30          0.44   
8            Enco         Žibrik         7.78       7.34          0.44   
9         Camilla       Zaitseva         7.53       7.13          0.40   
10          Tekla        Walczak         7.61       7.29          0.32   
11    Brancaleone          Russo         8.61       8.18          0.43   
12           Isac           Berg         9.68       9.29          0.39   
13        Clinton         Miller         7.79       7.40          0.39   
14         Eugene        Mironov         7.81       7.48          0.33   
15    Szilveszter          Totth         7.70       7.38          0.32   
16      Alexander      Mathiesen         7.96       7.55          0.41   
17            Chỉ            Lâm         7.68       7.24          0.44   
18      Władysław      Wieczorek         7.92       7.47          0.45   
19       Kristján        Ingason         7.92       7.57          0.35   
20         Marija       Grubišić         7.53       7.15          0.38   
21          Sauli     Koivuniemi         7.67       7.37          0.30   
22        Mariana          Souza         7.86       7.51          0.35   
23     Kristoffer      Martinsen         9.18       8.64          0.54   
24             Mỹ          Quynh         7.61       7.16          0.45   
25           Oles        Zhdanov         7.52       7.11          0.41   
26        Triana.       Terrazas         7.71       7.34          0.37   
27         Gabryŝ    Tomaszewski         7.87       7.47          0.40   
28      Leixandre         Alanis         7.74       7.32          0.42   
29   Onyekachukwu         Obinna         7.58       7.12          0.46   
..            ...            ...          ...        ...           ...   
320          Jane        Citizen         7.98       7.60          0.38   
321        Angela     Lavrentyev         7.61       7.14          0.47   
322        Edelma    Villalpando         7.99       7.56          0.43   
323        Annika          Vaara         7.73       7.34          0.39   
324         Chiho           Higa         7.71       7.30          0.41   
325     Beatrycze        Woźniak         7.54       7.17          0.37   
326      Miłosław     Wiśniewski         7.51       7.08          0.43   
327       Firenze          Fodor         7.89       7.55          0.34   
328           Zoe        Wellish         7.71       7.30          0.41   
329           Una  Traustadóttir         8.00       7.50          0.50   
330         Luboš          Pecha         7.79       7.45          0.34   
331         Meaza         Brhane         7.70       7.36          0.34   
332         Adlan       Shishani         7.84       7.37          0.47   
333         Sofia      Hermansen         8.90       8.57          0.33   
334         Guðni      Heimisson         7.64       7.24          0.40   
335       Eufemio        Rosario         7.54       7.26          0.28   
336      Dalmacia         Madrid         7.67       7.21          0.46   
337         Daimy          Tromp         9.41       8.94          0.47   
338        Jeremy        Montagu         7.68       7.36          0.32   
339       Nebechi      Ekechukwu         7.78       7.39          0.39   
340       Satsita      Batukayev         7.63       7.25          0.38   
341       Timothy         Cotton         7.92       7.52          0.40   
342       B

### What is Assessing Data?

The data analysis process consists of the following steps

1. Asking Questions
2. Data Wrangling<br><br>
    a. Gathering Data
            i.   CSV files
            ii.  APIs
            iii. Web Scraping
            iv.  Databases
    b. <b>Assessing Data</b><br>
    c. Cleaning Data<br><br>
3. Exploratory Data Analysis
4. Drawing Conclusion
5. Comunicating Results
    

### Assessing Data

In this step, the data is to be understood more deeply. Before implementing methods to clean it, you will definitely need to have a better idea about what the data is about.

### Types of Unclean Data

There are 2 kinds of unclean data

1. <u>Dirty Data</u> (Data with Quality issues):<br>
    Dirty data, also known as low quality data. Low quality data has content issues.<br>
2. <u>Messy Data</u> (Data with tidiness issues)<br>
    Messy data, also known as untidy data. Untidy data has structural issues.

### 1. Dirty Data

When the data has following kinds of issues:

1. Duplicated data
2. Missing Data
3. Corrupt Data
4. Inaccurate Data

This kind of data is known as Dirty data

### 2. Messy Data

When the data has structural issue is it known as Messy or untidy data.

Tidy data has the following properties:

1. Each variable forms a column
2. Each observation forms a row
3. Each observational unit forms a table

If a data violates any one of the above it is known as untidy data

Example:

<img src="untidy.png">

### Types of Assessment

There are 2 types of assessment styles

1. Manual - Looking through the data manually in google sheets
2. Programmatic - By using pandas functions such as info(), describe() or sample()

### Steps in Assessment

There are 2 steps involved in Assessment

1. Discover
2. Document

### Data Quality Dimensions

1. Completeness Issues
2. Validity Issues
3. Accuracy Issues
4. Consistency Issues